optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/common/Validator.java

```diff
@@ -0,0 +1,112 @@
+package org.optaplanner.openshift.employeerostering.gwtui.client.common;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Predicate;
+
+import com.google.gwt.event.dom.client.HasChangeHandlers;
+import com.google.gwt.event.dom.client.HasKeyUpHandlers;
+import com.google.gwt.event.logical.shared.HasValueChangeHandlers;
+import com.google.gwt.user.client.ui.HasEnabled;
+import com.google.gwt.user.client.ui.HasValue;
+
+public class Validator {
+
+    private HasEnabled submitButton;
+    private List<ValidatedInput<?>> toValidate;
+    private List<ValidationListener> validationListeners;
+    private boolean isValid;
+
+    public Validator(HasEnabled submitButton) {
+        this.submitButton = submitButton;
+        toValidate = new ArrayList<>();
+        validationListeners = new ArrayList<>();
+    }
+
+    public Validator assertNotEmpty(HasValue<String> input) {
+        return assertProperty((s) -> !s.trim().isEmpty(), input);
+    }
+
+    public Validator assertTextInRange(HasValue<String> input, int min, int max) {
+        return assertProperty((s) -> {
+            if (!CommonUtils.isInteger(s)) {
+                return false;
+            }
+            int value = Integer.parseInt(s);
+            return min <= value && value <= max;
+        }, input);
+    }
+
+    public Validator assertNumberInRange(HasValue<Integer> input, int min, int max) {
+        return assertProperty((value) -> {
+            return min <= value && value <= max;
+        }, input);
+    }
+
+    public <T> Validator assertProperty(Predicate<T> property, HasValue<T> input) {
+        toValidate.add(new ValidatedInput<T>(this, input, property));
+        update();
+        return this;
+    }
+
+    public Validator addValidationListener(ValidationListener listener) {
+        validationListeners.add(listener);
+        return this;
+    }
+
+    public void update() {
+        isValid = toValidate.stream().allMatch((i) -> i.isValid());
+        submitButton.setEnabled(isValid);
+    }
+
+    @SuppressWarnings("unchecked")
+    public void invalidate() {
+        toValidate.stream().forEach((i) -> ((ValidatedInput<Object>) i).onValueChange(i.element.getValue()));
+        update();
+    }
+
+    public void updateListeners(HasValue<?> element, boolean isValid) {
+        validationListeners.forEach((l) -> l.onValidationEvent(element, isValid));
+    }
+
+    public final static class ValidatedInput<T> {
+
+        private final HasValue<T> element;
+        private final Predicate<T> validator;
+        private final Validator parent;
+        private boolean isValid;
+
+        public ValidatedInput(Validator parent,
```

Comment:
All of the code you have for registering handlers is logic we have in Errai UI to deal with `@Bound` UI fields. Are you using this validation with Errai `@Templated` views? In that case, you might be better off using Errai data-binding and which comes with property change handlers.